contracts/TwabRewards.sol

```diff
@@ -0,0 +1,451 @@
+        returns (bool)
+    {
+        require(_isPromotionActive(_promotionId) == true, "TwabRewards/promotion-not-active");
+        require(_to != address(0), "TwabRewards/recipient-not-zero-address");
+
+        Promotion memory _promotion = _getPromotion(_promotionId);
+
+        IERC20 _token = IERC20(_promotion.token);
+        uint256 _remainingRewards = _getRemainingRewards(_promotionId);
+
+        if (_remainingRewards > 0) {
+            _token.safeTransfer(_to, _remainingRewards);
+        }
+
+        delete _promotions[_promotion.id];
+
+        emit PromotionCancelled(_token, _remainingRewards);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function extendPromotion(uint256 _promotionId, uint256 _numberOfEpochs)
+        external
+        override
+        onlyPromotionCreator(_promotionId)
+        returns (bool)
+    {
+        require(_isPromotionActive(_promotionId) == true, "TwabRewards/promotion-not-active");
+
+        Promotion memory _promotion = _getPromotion(_promotionId);
+        uint256 _extendedNumberOfEpochs = _promotion.numberOfEpochs + _numberOfEpochs;
+
+        _requireEpochLimit(_extendedNumberOfEpochs);
+
+        IERC20 _token = IERC20(_promotion.token);
+        uint256 _amount = _numberOfEpochs * _promotion.tokensPerEpoch;
+
+        _token.safeTransfer(address(this), _amount);
+        _promotions[_promotionId].numberOfEpochs = _extendedNumberOfEpochs;
+
+        emit PromotionExtended(_token, _amount, _extendedNumberOfEpochs);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getPromotion(uint256 _promotionId) external view override returns (Promotion memory) {
+        return _getPromotion(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getRemainingRewards(uint256 _promotionId) external view override returns (uint256) {
+        return _getRemainingRewards(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getCurrentEpoch(uint256 _promotionId) external view override returns (Epoch memory) {
+        return _getCurrentEpoch(_promotionId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getEpoch(uint256 _epochId, uint256 _promotionId)
+        external
+        view
+        override
+        returns (Epoch memory)
+    {
+        Promotion memory _promotion = _getPromotion(_promotionId);
+
+        return _getEpoch(_epochId, _promotion);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getRewardAmount(
+        address _user,
+        uint256 _promotionId,
+        uint256 _epochId
+    ) external view override returns (uint256) {
+        return _calculateRewardAmount(_user, _promotionId, _epochId);
+    }
+
+    /// @inheritdoc ITwabRewards
+    function claimRewards(
+        address _user,
+        uint256 _promotionId,
+        uint256 _epochId
+    ) external override returns (uint256) {
+        require(_epochId < _getCurrentEpoch(_promotionId).id, "TwabRewards/epoch-not-over");
```

Comment:
You see how having an Epoch struct actually makes this a little awkward?

Instead of a struct, better to have:

`getCurrentEpochId`

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/423f188e97b1edd4ff89b6b5621bde3f656b0fbc